robotiq_s_model_control/src/robotiq_s_model_control/s_model_can_client.cpp

```diff
@@ -0,0 +1,421 @@
+            read_mutex.unlock();
+            throw std::runtime_error(err.str().c_str());
+        }
+
+        read_mutex.unlock();
+    }
+    return input_;
+}
+
+SModelCanClient::GripperOutput SModelCanClient::readOutputs() const
+{
+    return output_;
+}
+
+void SModelCanClient::stateCallback(const can::State &s)
+{
+    std::string err;
+    driver_->translateError(s.internal_error, err);
+    if(!s.internal_error)
+    {
+        std::cout<<"State: " <<err <<", asio: "<<s.error_code.message()<<std::endl;
+        //throw std::runtime_error(s.error_code.message());
+    }
+    else
+    {
+        std::cout<<"Error: "<<err<<", asio: "<<s.error_code.message()<<std::endl;
+        throw std::runtime_error(err);
+    }
+
+}
+
+void SModelCanClient::frameCallback(const can::Frame &f)
+{
+    can::Frame frame = f;
+
+    if(!frame.isValid())
+    {
+        std::stringstream err;
+        err << "Invalid frame from SocketCAN: id " <<std::hex << f.id <<", length "<<(int) f.dlc <<", is_extended: "<<f.is_extended<<", is error: "<<f.is_error<<", is_rtr: "<<f.is_rtr;
+        throw std::runtime_error(err.str().c_str());
+    }
+    if (frame.is_error)
+    {
+        throw std::runtime_error("Received frame is error");
+    }
+    if ((int)frame.dlc != 8)
+    {
+        throw std::runtime_error("Message length must be 8");
+    }
+
+    if(f.data[0]==0x60)
+    {
+        //! received response from write request
+    }
+    else if (f.data[0]==0x4f && f.data[1]==0x00 && f.data[2] == 0x20)
+    {
+        switch((int)f.data[3])
+        {
+        case 0x01:
+            decodeGripperStatus(f.data[4]);
+            break;
+        case 0x02:
+            decodeObjectStatus(f.data[4]);
+            break;
+        case 0x03:
+            decodeFaultStatus(f.data[4]);
+            break;
+        case 0x04:
+            decodeFingerAPosCmd(f.data[4]);
+            break;
+        case 0x05:
+            decodeFingerAPos(f.data[4]);
+            break;
+        case 0x06:
+            decodeFingerACurrent(f.data[4]);
+            break;
+        case 0x07:
+            decodeFingerBPosCmd(f.data[4]);
+            break;
+        case 0x08:
+            decodeFingerBPos(f.data[4]);
+            break;
+        case 0x09:
+            decodeFingerBCurrent(f.data[4]);
+            break;
+        case 0x0A:
+            decodeFingerCPosCmd(f.data[4]);
+            break;
+        case 0x0B:
+            decodeFingerCPos(f.data[4]);
+            break;
+        case 0x0C:
+            decodeFingerCCurrent(f.data[4]);
+            break;
+        case 0x0D:
+            decodeFingerSPosCmd(f.data[4]);
+            break;
+        case 0x0E:
+            decodeFingerSPos(f.data[4]);
+            break;
+        case 0x0F:
+            decodeFingerSCurrent(f.data[4]);
+            break;
+        default:
+            //! unknown subindex
+            break;
+
+        }
+        read_mutex.unlock();
```

Comment:
The corresponding lock is on line 200, so that the function waits for the corresponding read reply before returning.  It times out and throws an error if the read reply is dropped for some reason.